[EP/DP][API Server] Enable DP-aware routing in OpenAI API requests

Prowindy opened the pull request:
Signed-off-by: Cong Chen <prowindy@gmail.com>
  ## Purpose

  This PR adds support for the `data_parallel_rank` parameter in vLLM's OpenAI API, enabling external routers to specify which data parallel rank
   should handle a request. This enhancement improves load distribution and cache locality in multi-GPU deployments by allowing DP-aware routing
  from router to engine core.

  **Key functionality:**
  - External routers can inject `data_parallel_rank` into API requests
  - vLLM server routes requests to specific GPU ranks based on the parameter
  - Maintains backward compatibility when parameter is not provided
  - Enables intelligent load balancing and cache-aware routing strategies

  ## Test Plan

  1. **End-to-End DP-Aware Routing Test**:
     ```bash
     # Start vLLM server with 8 DP ranks
     vllm serve meta-llama/Meta-Llama-3.1-8B-Instruct \
       --data-parallel-size 8 --tensor-parallel-size 1 --port 8000

     # Start vLLM-router with DP-aware routing (selective ranks 0,2,4,6)
     vllm-router --worker-urls http://0.0.0.0:8000 \
       --dp-aware --policy round_robin --port 30000

     # Generate test traffic
     vllm bench serve --num-prompts 100 --port 30000 \
       --endpoint /v1/completions --max-concurrency 32

  2. Validation Methods:
    - Added comprehensive logging throughout the request pipeline
    - Implemented rank counters to track request distribution
    - Verified GPU device assignment for each request
    - Tested with multiple routing policies (round_robin, cache_aware)

  Test Result

  ✅ Successful DP-Aware Routing Validation:

  1. Router Discovery: Successfully identified 4 DP-aware workers
  workers: ["http://0.0.0.0:8000@0", "http://0.0.0.0:8000@2", "http://0.0.0.0:8000@4", "http://0.0.0.0:8000@6"]
  2. Request Distribution: Balanced traffic across selected ranks
  Current counters: {0: 2, 2: 3, 4: 3, 6: 2}
  3. End-to-End Pipeline Verification:
  API_SERVER_DEBUG: Received completion request with data_parallel_rank=4
  ROUTING_DEBUG: Using data_parallel_rank=4 directly
  ENGINE_DISPATCH DEBUG: Request dispatched to engine identity 0400 (rank 4)
  GPU_DEVICE_DEBUG: Request received on GPU device, parallel_config.rank=4
  4. Multi-Process Validation: Confirmed different engine processes correctly received assigned requests:
    - EngineCore_0 pid=296341 → Rank 0 requests
    - EngineCore_2 pid=296343 → Rank 2 requests
    - EngineCore_4 pid=296345 → Rank 4 requests
    - EngineCore_6 pid=296347 → Rank 6 requests

Prowindy commented:
@robertgshaw2-redhat The current approach is consistent with how SGLang handles DP-aware requests. With this change to the vLLM server, the sgl-router fork can send requests directly to the vLLM server without requiring any API modifications.

That said, your suggestion is valid—P/D disaggregation also embeds the prefill/decode address in the X-Request-Id. I’ll make a few adjustments to incorporate your recommendation.

robertgshaw2-redhat commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -386,6 +386,24 @@ async def get_server_load_metrics(request: Request):
 
 
+
+@router.get("/get_server_info")
```
Is this needed? I think its usually not great to allow external users to see this type of server info

I think that instead we should make sure the server is in DEV mode to expose this

njhill replied:
I agree it's good to keep the discovery part separate (can figure out in future PR). The header changes in this one look good to me now.

robertgshaw2-redhat commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -217,51 +205,150 @@ async def build_async_engine_client_from_engine_args(
-            "V1 is enabled, but got --disable-frontend-multiprocessing. "
-            "To disable frontend multiprocessing, set VLLM_USE_V1=0."
-        )
+    if envs.VLLM_USE_V1:
```
we only support V1 now. this other stuff needs to be reverted

njhill reviewed (commented):
Thanks @Prowindy! LGTM

Looks like there is a commit which needs DCO sign-off, see https://github.com/vllm-project/vllm/pull/24945/checks?check_run_id=53896850372, and some pre-commit errors.

Also I wonder if we could add something to the docs to cover this. Perhaps add a note to the external router section of the data parallel page: https://docs.vllm.ai/en/latest/serving/data_parallel_deployment.html#external-load-balancing

Prowindy committed:
Add data_parallel_rank support to OpenAI API for DP-aware routing

Signed-off-by: Cong Chen <prowindy@gmail.com>

This commit adds support for data parallel rank routing in the OpenAI API
to enable intelligent load balancing across data parallel instances.

Key changes:
- Add _get_data_parallel_rank() method to extract X-data-parallel-rank header
- Pass data_parallel_rank to engine.generate() in chat and completion APIs
- Add /get_server_info endpoint to return DP size for router discovery
- Add test coverage for data_parallel_rank extraction

The X-data-parallel-rank header allows external routers to specify which
data parallel rank should handle a request, enabling DP-aware request
routing and load balancing.
